### ISSUE TITLE
Sqlfluff doesn't support parameter markers

### ISSUE THREAD

**Issue by YilangHe** [sqlfluff/sqlfluff#4008]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

sqlfluff can not successfully parse parameter markers. For example, in the snowflake dialect, it cannot parse the sign `?`

### Expected Behaviour

parse `?` as parameter markers in snowflake dialect

In the future, support parameters for some other dialects
- ? (snowflake query history)
- $x ([snowflake](https://docs.snowflake.com/en/sql-reference/session-variables.html))
- :x ([sqlalchemy](https://stackoverflow.com/questions/19314342/python-sqlalchemy-pass-parameters-in-connection-execute)) 
- %s ([MySQL, pgSQL](https://security.openstack.org/guidelines/dg_parameterize-database-queries.html))

### Observed Behaviour

Get the parsing error with the parameter marks in the SQL query:

```shell
[L:  1, P:  1]      |file:
[L:  1, P:  1]      |    statement:
[L:  1, P:  1]      |        select_statement:
[L:  1, P:  1]      |            select_clause:
[L:  1, P:  1]      |                keyword:                                      'SELECT'
[L:  1, P:  7]      |                [META] indent:
[L:  1, P:  7]      |                whitespace:                                   ' '
[L:  1, P:  8]      |                select_clause_element:
[L:  1, P:  8]      |                    wildcard_expression:
[L:  1, P:  8]      |                        wildcard_identifier:
[L:  1, P:  8]      |                            star:                             '*'
[L:  1, P:  9]      |            whitespace:                                       ' '
[L:  1, P: 10]      |            [META] dedent:
[L:  1, P: 10]      |            from_clause:
[L:  1, P: 10]      |                keyword:                                      'FROM'
[L:  1, P: 14]      |                whitespace:                                   ' '
[L:  1, P: 15]      |                from_expression:
[L:  1, P: 15]      |                    [META] indent:
[L:  1, P: 15]      |                    from_expression_element:
[L:  1, P: 15]      |                        table_expression:
[L:  1, P: 15]      |                            table_reference:
[L:  1, P: 15]      |                                naked_identifier:             'T'
[L:  1, P: 16]      |                    [META] dedent:
[L:  1, P: 16]      |            whitespace:                                       ' '
[L:  1, P: 17]      |            where_clause:
[L:  1, P: 17]      |                unparsable:                                   !! Expected: 'where_clause'
[L:  1, P: 17]      |                    keyword:                                  'WHERE'
[L:  1, P: 22]      |                    whitespace:                               ' '
[L:  1, P: 23]      |                    raw:                                      '?'
[L:  1, P: 24]      |                    whitespace:                               ' '
[L:  1, P: 25]      |                    raw:                                      '>'
[L:  1, P: 26]      |                    whitespace:                               ' '
[L:  1, P: 27]      |                    raw:                                      '?'
[L:  1, P: 28]      |    [META] end_of_file:

==== parsing violations ====
L:   1 | P:  17 |  PRS | Line 1, Position 17: Found unparsable section: 'WHERE ? > ?'
WARNING: Parsing errors found and dialect is set to 'snowflake'. Have you configured your dialect correctly?

Process finished with exit code 1


```

### How to reproduce

```sql
SELECT * FROM T WHERE COL > ?
```

### Dialect

snowflake

### Version

sqlfluff, version 1.3.2

Python 3.10.7

### Configuration

[sqlfluff]
dialect = snowflake
templater = raw

[sqlfluff:rules:L057]
additional_allowed_characters = $

### Are you willing to work on and submit a PR to address the issue?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by barrywhart** [comment]
This is not necessary. See comment on the related PR. Closing.
